### DISCUSSION TITLE
Deployer 7 - Labels, stages and roles

### DISCUSSION THREAD

**Discussion by franck-grenier** [deployphp/deployer#2766] [Help needed]
Hello Anton,
I'm using your great Deployer tool and I'm migrating from 6 to 7.

I need explanations about the "labels" concept introduced in Deployer 7 and how I can use it to declare and select stages and roles.

Please, have a look at my `deploy.php` file (version 7) below. Some tasks need to be run as `root` and others as `myuser`  : 

```php
host('dev-app')
    ->set('hostname', 'dev.mydomain.com')
    ->set('labels', [
        'stage' => 'dev',
        'role'  => 'app',
    ])
    ->set('remote_user', 'myuser');

host('dev-root')
    ->set('hostname', 'dev.mydomain.com')
    ->set('labels', [
        'stage' => 'dev',
        'role'  => 'root',
    ])
    ->set('remote_user', 'root');
  
task('server:install', function () {
    run('apt-get update');
    run('apt-get install acl');
})->select('role=root');

task('deploy:copy_conf', function () {
    $stage = input()->getArgument('stage');
    run("cd {{release_path}} && cp -r {$stage}/nginx/conf.d/* nginx/conf.d/");
});
```

How can I obtain the role `root` when I launch commands ? 

Is it done automatically by selecting the right host ?

`dep server:install dev-root` 

Or do I need to pass an option/argument in the command ? And what about "stages" ? How do they work with roles ?

Sorry, I'm feeling a bit lost with Deployer 7... I think we really need the documentation.

One more question : how can I get the `stage` argument value ?

```php
input()->getArgument('stage');
```

Throws : 

```
Symfony\Component\Console\Exception\InvalidArgumentException  in Input.php on line 110: The "stage" argument does not exist.
```

Thanks for your help.

**Comment by jonaswebdev** [comment]
About getting "stage" argument, I found it on [documentation](https://deployer.org/docs/7.x/selector), should help you :)

`    writeln('type:' . get('labels')['type'] . ' env:' . get('labels')['env']);`
